harlow/utils/helper_functions.py

```diff
@@ -2,64 +2,68 @@
 Helper functions for the adaptive sampling strategies.
 """
 
-import math
-
 import numpy as np
 import tensorflow as tf
 import tensorflow_probability as tfp
-from sklearn.metrics import mean_absolute_error, mean_squared_error
 from skopt.sampler import Lhs
 from skopt.space import Space
 
 tfd = tfp.distributions
 
 
-# TODO RRSE code should be checked for correctness
-def _rrse(x, y):
-    dims = x.shape
-    x_bar = np.mean(y)
-    x_bar_arr = np.zeros(dims)
-    x_bar_arr.fill(x_bar)
-
-    return math.sqrt(mean_squared_error(x, y) / mean_squared_error(x, x_bar_arr))
-
-
-def evaluate(metric, model, test_points_X, test_points_y):
+def evaluate_modellist(metric, model, test_points_X, test_points_y):
     """
     Evaluate user specified metric for the current iteration
 
     Returns:
     """
-    score_mtrx = np.zeros(len(np.atleast_1d(metric)))
-    count = 0
+    count_model = 0
+    metric_dict = {}
+    if not isinstance(metric, list):
+        metric = [metric]
     if metric is None or test_points_X is None:
-        score_mtrx[count] = 0.0
+        raise ValueError
     else:
-        for metric_func in np.atleast_1d(metric):
-            score_mtrx[count] = metric_func(model.predict(test_points_X), test_points_y)
-            count += 1
+        for metric_fun in metric:
+            scores = []
+            for m in model:
+                scores.append(
+                    metric_fun(m.predict(test_points_X), test_points_y[:, count_model])
+                )
+            metric_dict[metric_fun.__name__] = scores
 
-    return score_mtrx
+    return metric_dict
 
 
-def rrse(actual: np.ndarray, predicted: np.ndarray):
-    """Root Relative Squared Error"""
-    return np.sqrt(
-        np.sum(np.square(actual - predicted))
-        / np.sum(np.square(actual - np.mean(actual)))
-    )
+def evaluate(metric, true_y, predicted_y):
+    """
+    Evaluate user specified metric for the current iteration
 
+    Returns:
+    """
+    count_model = 0
+    metric_dict = {}
+    if not isinstance(metric, list):
+        metric = [metric]
+    if metric is None or true_y is None:
```

Comment:
Hi Merijn, I briefly looked over the integration test results. It seems that the change in the `evaluate` function in line 48 of `harlow/utils/helper_functions.py` leads to errors in the pipeline. Previously the `evaluate` function would return zero if no test points or metric was provided. Now it will instead throw an error.

Since there may be situations where one would want to run a sampler without providing test points, I think it would be preferable to refine this part. One possible solution would be to consider the different cases:

1. No metric and no test points -> Skip evaluation of metric
2. Metric without test points -> Evaluate the metric and throw error if that metric requires test points (assuming there are metrics we have not implemented that do not need test points, e.g. metrics based on the variance of the prediction for GP surrogates).
3. Test points without metric -> Error
4. Test points and metric -> Evaluate metric for test points

This should probably be solved in a different merge request (I could look into this as well), so I am fine with merging this branch.

Reply:
Hi Jan,

Thanks for spotting this. I completely agree with you that we should handle this according to the scheme you present.
I'll merge this branch for now and leave it as an issue we can pick up later :-)